### ISSUE TITLE
Is there any way to interrupt the current query of a pooled connection?

### ISSUE THREAD

**Issue by LucianZz** [boostorg/mysql#343]
Is there any way to interrupt the current query of a pooled connection?

**Comment by LucianZz** [comment]
Thank you for your reply. My need is to immediately interrupt the current query of a pooled connection, so my current approach is to obtain the CONNECTION_ID through "SELECT CONNECTION_ID()", and then interrupt the query through "KILL QUERY"

**Comment by anarthal** [comment]
That's an option, but you will need a separate connection to issue the `KILL QUERY` statement, as the server processes queries in the same connection sequentially. According to MySQL documentation, using this approach will leave your original connection intact.

You need to obtain the connection ID before issuing the query, then pass the ID to the connection you're using to issue the `KILL QUERY` command. In principle, this should work. Without knowing your application architecture, these are some problems I can think of:

- You want to interrupt a query, but all connections are busy running heavyweight queries, so your cancellation ends up being delivered late.
- Your cancellation is delivered after the query finishes, and ends up interrupting the next query.

I can't evaluate whether these are relevant or not without knowing more.

The alternative would be using a `parallel_group` with a second task that waits for some sort of cancellation signal to arrive, like this:

```cpp
// Task handling cancellations
asio::awaitable<void> wait_for_cancellation()
{
    // Wait for a cancellation signal to arrive. This depends on how
    // your application knows that the query should be cancelled -
    // maybe reading from a socket, receiving an HTTP request, a signal from a parent process...
    co_await something_specific_to_your_app();
}

asio::awaitable<void> interruptable_query(mysql::pooled_connection& conn)
{
    // Required to run async_execute
    mysql::results r;

    // Launch the query and the task that listens to the cancellation, in parallel. wait_for_one() will 
    // wait for the operation that completes first, and then cancel the other one
    // (similar to Promise.race in Javascript).
    // compl_order is a std::array<std::size_t, 2> with the order in with the ops completed.
    //     execute completed first: {0, 1}
    //     wait_for_cancellation wait completed first: {1, 0}
    // execute_err is the error code produced by the query
    // wait_for_cancellation_err is a std::exception_ptr produced by wait_for_cancellation
    auto [compl_order, execute_err, wait_for_cancellation_err] = co_await asio::experimental::make_parallel_group(
        conn->async_execute("SELECT 1", r, asio::deferred),
        asio::co_spawn(
            co_await asio::this_coro::executor,
            &wait_for_cancellation,
            asio::deferred
        )
    ).async_wait(asio::experimental::wait_for_one(), asio::deferred);

    // Check whether the operation completed successfully
    if (execute_err)
    {
        // We have an error. asio::error::operation_aborted means that the query was cancelled
    }
}
```

The viability of this approach depends on whether writing `something_specific_to_your_app` is doable or not.

Let me know if I can help more.

Regards,
Ruben.

**Comment by anarthal** [comment]
If you're okay with it, I will be closing this issue now. If you have any further questions, please feel free to open a new one.
